File: Encapsulation.py
class Student:
    def __init__(self):
        self.__marks = 0  # private

    def set_marks(self, m):   # setter
        if m >= 0 and m <= 100:
            self.__marks = m
        else:
            print("Invalid marks")

    def get_marks(self):      # getter
        return self.__marks
    
class King(Student):
    pass

s = Student()
s.set_marks(90)         # ✅ safe
print(s.get_marks())    # ✅ 90
k = King()
k.set_marks(30)
print(k.get_marks())


# print(s.__marks)      ❌ error: private

File: oops/Encapsulation.cpp
#include <iostream>
using namespace std;

class Student {
private:
    int marks;   // private variable

public:
    // Constructor
    Student() {
        marks = 0;   // default
    }

    // Setter
    void setMarks(int m) {
        if (m >= 0 && m <= 100) {
            marks = m;
        } else {
            cout << "Invalid marks" << endl;
        }
    }

    // Getter
    int getMarks() {
        return marks;
    }
};

// Derived class King inherits Student
class King : public Student {
    // nothing extra, inherits everything
};

int main() {
    Student s;
    s.setMarks(90);
    cout << s.getMarks() << endl;

    King k;
    k.setMarks(30);
    cout << k.getMarks() << endl;

    // cout << s.marks; ❌ ERROR: 'marks' is private in Student

    return 0;
}
